File: Robotics Kinematics and Dynamics Course/Chapter17/Lesson2/Chapter17_Lesson2_Ex1.cpp
#include <pinocchio/algorithm/jacobian.hpp>

pinocchio::Model model;
pinocchio::buildModels::humanoidRandom(model);
pinocchio::Data data(model);

Eigen::VectorXd q = pinocchio::randomConfiguration(model);
Eigen::VectorXd v = Eigen::VectorXd::Zero(model.nv);

// forward kinematics
pinocchio::forwardKinematics(model, data, q, v);

// Jacobian of a given frame
pinocchio::FrameIndex fid = model.getFrameId("left_foot");
Eigen::Matrix<double,6,Eigen::Dynamic> J6;
pinocchio::computeFrameJacobian(model, data, q, fid,
                                pinocchio::LOCAL_WORLD_ALIGNED, J6);
      

File: Robotics_Kinematics_and_Dynamics_Course/Chapter17/Lesson2/Chapter17_Lesson2_Ex1.py
import pinocchio as pin

# Load model from URDF
model = pin.buildSampleModelHumanoidRandom()
data = model.createData()

# q: configuration vector including floating base (dimension model.nq)
# v: velocity vector including base twist (dimension model.nv)
q = pin.randomConfiguration(model)
v = np.zeros(model.nv)

# Forward kinematics and Jacobians in one call
pin.forwardKinematics(model, data, q, v)
frame_id = model.getFrameId("left_foot")
J6 = pin.computeFrameJacobian(model, data, q, frame_id, pin.ReferenceFrame.LOCAL_WORLD_ALIGNED)
      
